pkg/util/pki/csr.go

```diff
@@ -341,7 +345,61 @@ func GenerateCSR(crt *v1.Certificate, optFuncs ...GenerateCSROption) (*x509.Cert
 // key of the signer.
 // It returns a PEM encoded copy of the Certificate as well as a *x509.Certificate
 // which can be used for reading the encoded values.
-func SignCertificate(template *x509.Certificate, issuerCert *x509.Certificate, publicKey crypto.PublicKey, signerKey interface{}) ([]byte, *x509.Certificate, error) {
+func SignCertificate(template *x509.Certificate, issuerCert *x509.Certificate, publicKey crypto.PublicKey, signerKey any) ([]byte, *x509.Certificate, error) {
+	typedSigner, ok := signerKey.(crypto.Signer)
+	if !ok {
+		return nil, nil, fmt.Errorf("didn't get an expected Signer in call to SignCertificate")
+	}
+
+	var signatureAlgorithm x509.SignatureAlgorithm
+
+	// NB: can't rely on issuerCert.Public or issuercert.PublicKeyAlgorithm being set reliably;
```

Comment:
We already have a SignatureAlgorithm function which we might be able to use here instead: https://github.com/cert-manager/cert-manager/blob/1be0dfc2e96ec43617fd49dd88b1b02422f544a8/pkg/util/pki/csr.go#L440

Reply:
Yeah... I saw it and obviously `v1.Certificate` isn't really helpful to use from `SignCertificate`. It feels wrong to create a dummy `v1.Certificate` just to be able to use this. I could refactor both my new changes and this function to use a different API though.

Reply:
Ah yes, that looks complex, maybe we just add a comment to each function to make readers aware of the other function? Also, one is used to determine the CSR signature and the other for the certificate signature. For some reason, I find this code very interesting to read, very fun.

Reply:
Have refactored into a separate func which SignatureAlgorithm and SignCertificate now call!